### PR TITLE
Improve docs of the Warning module

### PR DESCRIPTION
**Pull request by eregon**

### CODE REVIEW

**Review by jeremyevans — APPROVED** [review]
Looks good!

File: error.c [hunk 1]
```diff
@@ -265,11 +267,30 @@ rb_warning_s_warn(int argc, VALUE *argv, VALUE mod)
...
+ *  handled by Ruby, for instance by filtering some warnings, and/or outputting
+ *  warnings somewhere other than $stderr.
+ *
+ *  If you want to change the behavior of Warning.warn you should use
+ *  +Warning.extend(MyNewModuleWithWarnMethod)+ and you can use `super`
+ *  to get the default behavior of printing the warning to $stderr.
+ *
+ *  Example:
+ *    module MyWarningFilter
+ *      def warn(message)
```
Comment (eregon) [inline error.c:280]: @jeremyevans Do you think we should show the `category` keyword or it's fine like this?
It looks a bit more complicated:
```ruby
  def warn(message, category: nil, **kwargs)
# or
  def warn(message, **kwargs)
```

**Comment by jeremyevans** [comment]
It should be `def warn(message, category: nil)`.  I fixed `Warning.warn` to only accept the `:category` keyword.

**Comment by eregon** [comment]
What if we add new keywords in the future?
Then it's an ArgumentError on purpose to signal it does not support it yet?
Might get complicated if there are multiple `extend`'s of Warning.warn and how e.g. `Kernel#warn` should call it with more or less kwargs.
`def warn(message, category: nil, **kwargs)` seems more future-proof to me (worse case: it ignores extra additional info).

**Comment by jeremyevans** [comment]
It's a tradeoff.  If the method override always calls `super`, and does not use the category, then just `**kwargs` is fine.  If it uses the category, then `category: nil, **kwargs` is fine.  If the method override does not always call `super`, then using `**kwargs` you can end up with cases where bad keywords passed do not result in an error. I have no strong feelings about this, it's your call.

**Comment by eregon** [comment]
OK, I updated to `def warn(message, category: nil, **kwargs)` for this example.

**Pull request merged**